### PR TITLE
cleaned up doxygen for g2 lib on a, c, d, e named code files

### PR DESCRIPTION
**Pull request by Hang-Lei-NOAA**
Fix the issues on pull request  add doxygen for g2 lib on a, c, d, e named code files #66
reopen it from a feature branch.
https://hang-lei-noaa.github.io/NCEPLIBS/NCEPLIBS-g2/html/files.html

### CODE REVIEW

File: src/addlocal.f [hunk 1]
```diff
@@ -1,45 +1,33 @@
+!>    @file
+!>    @brief This subroutine adds a Local Use Section (Section 2) to
+!>    a GRIB2 message.
+!>    @author Gilbert ORG: W/NP11 @date 2000-05-01
+!>    
+
+!>    This routine is used with routines gribcreate(), addlocal(), addfield(),
```
Comment (edwardhartnett) [inline src/addlocal.f:8]: The same info in the @brief above needs to be here also. So change this to:

```
!> This subroutine adds a Local Use Section (Section 2) to
!>    a GRIB2 message. This routine is used with routines gribcreate(), addlocal(), addfield(), ...
```

In the doxygen output, this information appears in two different places. So it needs to be both in the file @brief, and as the first sentence of the subroutine documentation. This applies to all the files you doxygenate.

File: src/comunpack.f [hunk 2]
```diff
@@ -1,47 +1,37 @@
...
+!>    complex packing algorithm as defined in the GRIB2 documention.
+!>    @author Gilbert ORG: W/NP11 @date 2000-06-21
+!>     
+
+!>    Using info from the GRIB2 Data Representation Template 5.2 or 5.3.
+!>    Supports GRIB2 complex packing templates with or without
+!>    spatial differences (i.e. DRTs 5.2 and 5.3).
+!>    
+!>    PROGRAM HISTORY LOG:
+!>    2000-06-21  Gilbert
```
Comment (edwardhartnett) [inline src/comunpack.f]: In order to appear as a list in doxygen, there needs to be a hyphen at the beginning of each line. So:

!>    PROGRAM HISTORY LOG:
!>    - 2000-06-21  Gilbert
!>    - 2004-12-29  Gilbert  -  Added test ( provided by Arthur Taylor/MDL )
!>    to verify that group widths and lengths are consistent with section length.
!>    - 2016-02-26              update unpacking for template 5.3

File: src/dec_jpeg2000.c [hunk 3]
```diff
@@ -24,46 +31,24 @@
...
+ * The JPEG2000 Part-1 standard (i.e., ISO/IEC 15444-1) using JasPer 
+ * Software version 1.500.4 (or 1.700.2) written by the University of British
+ * Columbia and Image Power Inc, and others.
+ * JasPer is available at http://www.ece.uvic.ca/~mdadams/jasper/.
+ *    
+ * @param[in] injpc Input JPEG2000 code stream.
+ * @param[in] bufsize Length (in bytes) of the input JPEG2000 code stream.
+ * @param[in] outfld Output matrix of grayscale image values. Only grayscale is expected.
+ * @return - > 0 = Successful decode
+ *          - -3 = Error decode jpeg2000 code stream.
```
Comment (edwardhartnett) [inline src/dec_jpeg2000.c]: The hyphens need to line up vertically to be properly interpreted.

File: src/dec_png.c [hunk 4]
```diff
@@ -1,3 +1,8 @@
+/**
+ * @file
+ * @brief This struct read and decode PNG stream
```
Comment (edwardhartnett) [inline src/dec_png.c]: This needs to be a properly punctuated sentence. 

Also there need to be an author tag for the file.

File: src/dec_png.c [hunk 5]
```diff
@@ -148,7 +176,9 @@ int SUB_NAME(unsigned char *pngbuf,g2int *width,g2int *height,char *cout)
       }
     }
 
-/*      Clean up   */
+/*
+**     Clean up
+*/
```
Comment (edwardhartnett) [inline src/dec_png.c:182]: For the rest of this doxygenation task, please put up each file as a separate PR, to make them easier to review, and please don't reformat the comments in the code. If you want to do that in the future, you can, but we can discuss what the format should be. For now, let's just focus on doxygenation.

File: src/drstemplates.f [hunk 6]
```diff
@@ -56,96 +45,83 @@ module drstemplates
 
       type(drstemplate),dimension(MAXTEMP) :: templates
 
-      data templates(1)%template_num /0/     !  Simple Packing
+      data templates(1)%template_num /0/     !<     Simple Packing
```
Comment (edwardhartnett) [inline src/drstemplates.f:50]: These should not be made into doxygen comments.

File: src/drstemplates.f [hunk 7]
```diff
@@ -56,96 +45,83 @@ module drstemplates
...
 
 
       contains
 
+!>    @brief  This function returns the index of specified Data
+!>    Representation Template 5.NN (NN=number) in array templates.                                    .
+!>    @author Gilbert         ORG: W/NP11    DATE: 2001-06-28
+!>     
+
+!>    @param[in] number NN, indicating the number of the Data Representation
```
Comment (edwardhartnett) [inline src/drstemplates.f:118]: There needs to be a first sentence describing the function before the @param tags.

File: src/drstemplates.f [hunk 8]
```diff
@@ -56,96 +45,83 @@ module drstemplates
...
+!         data templates(5)%needext /.true./
+!         data (templates(5)%mapdrs(j),j=1,15)
+!        &                        /4,-2,-2,1,1,1,4,2,2,1,1,1,1,1,1/
 
 
       contains
 
+!>    @brief  This function returns the index of specified Data
+!>    Representation Template 5.NN (NN=number) in array templates.                                    .
+!>    @author Gilbert         ORG: W/NP11    DATE: 2001-06-28
```
Comment (edwardhartnett) [inline src/drstemplates.f]: THe @date tag is misisng.

File: src/drstemplates.f [hunk 9]
```diff
@@ -56,96 +45,83 @@ module drstemplates
...
+!>     
+
+!>    @param[in] number NN, indicating the number of the Data Representation 
+!>    Template 5.NN that is being requested.
+!>
+!>    @return - > Index of DRT 5.NN in array templates, if template exists.
+!>    = -1, otherwise.
+!>
+!>    @author Gilbert         ORG: W/NP11    DATE: 2001-06-28
```
Comment (edwardhartnett) [inline src/drstemplates.f]: @date tag missing.

File: src/drstemplates.f [hunk 10]
```diff
@@ -159,44 +135,29 @@ integer function getdrsindex(number)
...
          end function
 
-
+!>    @brief This subroutine returns DRS template information for a                                   .
+!>    specified Data Representation Template 5.NN.
+!>    @author Gilbert         ORG: W/NP11    DATE: 2000-05-11
+!>     
+
+!>    The number of entries in the template is returned along with a map
+!>    of the number of octets occupied by each entry.  Also, a flag is
```
Comment (edwardhartnett) [inline src/drstemplates.f]: Remove the nonsense of two spaces after periods. Not necessary with doxygen.

File: src/drstemplates.f [hunk 11]
```diff
@@ -219,38 +180,26 @@ subroutine getdrstemplate(number,nummap,map,needext,iret)
 
          end subroutine
 
+!>    @brief This subroutine generates the remaining octet map for a
+!>    given Data Representation Template, if required.
+!>    @author  Gilbert         ORG: W/NP11    DATE: 2000-05-11
```
Comment (edwardhartnett) [inline src/drstemplates.f]: @date tag.

File: src/enc_jpeg2000.c [hunk 12]
```diff
@@ -23,65 +30,42 @@
    #define SUB_NAME enc_jpeg2000_
 #endif
 
+/**
+ * Using JasPer Software version 1.500.4 (or 1.700.2 ) written by the 
+ * University of British Columbia, Image Power Inc, and others.
+ * JasPer is available at http: *   www.ece.uvic.ca/~mdadams/jasper/.
+ *  
+ * PROGRAM HISTORY LOG:
+ * 2002-12-02  Gilbert
```
Comment (edwardhartnett) [inline src/enc_jpeg2000.c]: Format as list.

**Review by edwardhartnett — CHANGES_REQUESTED** [review]
Let's try again but just one file at a time, so I don't have to keep adding the same comments to each file.

When you get to the point where you are doxygenating files without any review comments, then you can try multiple files at once. ;-)

File: src/dec_png.c (hunk 5 above)
Comment (Hang-Lei-NOAA) [reply src/dec_png.c:182]: Okay

**Comment by edwardhartnett** [comment]
Also please use better titles for PRs. See mine for examples.

**Comment by Hang-Lei-NOAA** [comment]
@edwardhartnett 
Could you please review the changes? Thanks

File: src/cmplxpack.f [hunk 13]
```diff
@@ -1,59 +1,34 @@
...
+!>    @param[in] idrsnum Data Representation Template number 5.N Must equal 2 or 3.
+!>    @param[inout] idrstmpl Contains the array of values for Data Representation Template 5.2 or 5.3
+!>    - (1) = Reference value - ignored on input set by compack routine.
+!>    - (2) = Binary Scale Factor
+!>    - (3) = Decimal Scale Factor
+!>    - (7) = Missing value management
+!>    - (8) = Primary missing value
+!>    - (9) = Secondary missing value
+!>    - (17) = Order of Spatial Differencing  ( 1 or 2 )
+!>    @param[out] pack The packed data field (character*1 array)
```
Comment (GeorgeGayno-NOAA) [inline src/cmplxpack.f:25]: This should be 'cpack', not 'pack'.

Comment (GeorgeGayno-NOAA) [reply src/cmplxpack.f:25]: When I compiled, doxygen gave me this warning: 
_warning: argument 'pack' of command @param is not found in the argument list...._

It is good practice to check the compilation log for warnings.

**Review by edwardhartnett — APPROVED** [review]

**Review by GeorgeGayno-NOAA — COMMENTED** [review]
I don't know why my comments did not show up. Can you address this during your next pull request.

**Comment by Hang-Lei-NOAA** [comment]
Will do.
Thanks,
Hang

**Pull request merged**